File: gfssoc1j5.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class gfssoc1j5 {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;

	static int nextInt() throws IOException {
		while (st == null || !st.hasMoreTokens()) {
			st = new StringTokenizer(br.readLine().trim());
		}
		return Integer.parseInt(st.nextToken());
	}

	public static void main(String[] args) throws IOException {
		ArrayList<Integer>[] adj = new ArrayList[nextInt()+1];

		int times = nextInt();
		int minutes = nextInt();

		for (int i = 0; i < adj.length; i++) {
			adj[i] = new ArrayList<Integer>();
		}
		
		for (int i = 0; i < times; i++) {
			adj[nextInt()].add(nextInt());
		}

		int queries = nextInt();

		for (int i = 0; i < queries; i++) {
			int start = nextInt();
			int end = nextInt();
			bfs(adj, start, end, minutes);
		}
	}

	static void bfs(ArrayList<Integer>[] adj, int start, int end, int minutes) {
		Queue<Integer> q = new LinkedList<Integer>();
		boolean[] visited = new boolean[adj.length];
		int[] distance = new int[adj.length];

		visited[start] = true;
		q.add(start);

		boolean found = false;
		while (!q.isEmpty()) {
			int x = q.poll();

			for (int i : adj[x]) {
				if (!visited[i]) {
					visited[i] = true;
					distance[i] = distance[x] + 1;
					if (i == end) {
						found = true;
						break;
					}
					q.add(i);
				}
			}

			if (found) {
				break;
			}
		}
		System.out.println(found ? distance[end] * minutes : "Not enough hallways!");
	}
}

File: gfssoc1j5.cpp
#include <cstring>
#include <iostream>
#include <queue>
#include <vector>

void bfs(std::vector<int> *adj, int start, int end, int minutes) {
	std::queue<int> q;
	bool visited[1001];
	int distance[1001];

	std::memset(distance, 0, sizeof(distance));
	std::memset(visited, 0, sizeof(visited));

	visited[start] = true;
	q.push(start);

	bool found = false;
	while (!q.empty()) {
		int x = q.front();
		q.pop();
		for (int i : adj[x]) {
			if (!visited[i]) {
				visited[i] = true;
				distance[i] = distance[x] + 1;
				if (i == end) {
					found = true;
					break;
				}
				q.push(i);
			}
		}
		if (found) break;
	}

	if (found) {
		std::cout << distance[end] * minutes << '\n';
	} else {
		std::cout << "Not enough hallways!" << '\n';
	}
}

int main() {
	std::cin.sync_with_stdio(0);
	std::cin.tie(0);

	std::vector<int> adj[1001];
	int rooms, times, minutes; std::cin >> rooms >> times >> minutes;

	for (int i = 0, x, y; i < times; i++) {
		std::cin >> x >> y;
		adj[x].push_back(y);
	}

	int queries; std::cin >> queries;

	for (int i = 0, start, end; i < queries; i++) {
		std::cin >> start >> end;
		bfs(adj, start, end, minutes);
	}
	return 0;
}